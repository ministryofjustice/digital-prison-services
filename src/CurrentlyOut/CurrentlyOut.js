import React from 'react'
import PropTypes from 'prop-types'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import SortableColumn from '../tablesorting/SortableColumn'
import OffenderName from '../OffenderName'
import DateFormatter from '../DateFormatter'
import OffenderLink from '../OffenderLink'
import OffenderImage from '../OffenderImage'
import SortLov from '../tablesorting/SortLov'
import Location from '../Location'
import { LAST_NAME } from '../tablesorting/sortColumns'
import Flags from '../FullFlags/Flags'
import './CurrentlyOut.scss'

const CurrentlyOut = ({ rows, sortOrder, setColumnSort }) => (
  <>
    <GridRow>
      <GridCol setWidth="one-quarter">
        <SortLov sortColumns={[LAST_NAME]} sortColumn={LAST_NAME} sortOrder={sortOrder} setColumnSort={setColumnSort} />
      </GridCol>
    </GridRow>
    <table className="currently-out">
      <thead>
        <tr>
          {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
          <th className="straight width10" />
          <th className="straight width10">
            <SortableColumn
              heading="Name"
              column={LAST_NAME}
              sortOrder={sortOrder}
              setColumnSort={setColumnSort}
              sortColumn={LAST_NAME}
            />
          </th>
          <th className="straight width10">Prison no.</th>
          <th className="straight width10">D.O.B.</th>
          <th className="straight width10">Location</th>
          <th className="straight width10">Incentive Level</th>
          <th className="straight width15">Flags</th>
          <th className="straight width10">Current location</th>
          <th className="straight width15">Comment</th>
        </tr>
      </thead>
      <tbody>
        {rows.map(row => (
          <tr className="row-gutters" key={row.offenderNo}>
            <td className="row-gutters">
              <OffenderLink offenderNo={row.offenderNo}>
                <OffenderImage offenderNo={row.offenderNo} />
              </OffenderLink>
            </td>
            <td className="row-gutters">
              <OffenderLink offenderNo={row.offenderNo}>
                <OffenderName lastName={row.lastName} firstName={row.firstName} />
              </OffenderLink>
            </td>
            <td className="row-gutters">{row.offenderNo}</td>
            <td className="row-gutters">
              <DateFormatter isoDate={row.dateOfBirth} />
            </td>
            <td className="row-gutters">
              <Location location={row.location} />
            </td>
            <td className="row-gutters">{row.iepLevel}</td>
            <td className="row-gutters">
              <Flags offenderNo={row.offenderNo} alerts={row.alerts} category={row.category} />
            </td>
            <td className="row-gutters">{row.toAgencyDescription || row.toCity}</td>
            <td className="row-gutters">{row.commentText}</td>
          </tr>
        ))}
      </tbody>
    </table>
  </>
)

CurrentlyOut.propTypes = {
  rows: PropTypes.arrayOf(
    PropTypes.shape({
      offenderNo: PropTypes.string.isRequired,
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
      location: PropTypes.string.isRequired,
      movementTime: PropTypes.string,
      fromAgencyDescription: PropTypes.string,
      alerts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
      category: PropTypes.string,
      toCity: PropTypes.string,
    })
  ),
  sortOrder: PropTypes.string,
  setColumnSort: PropTypes.func,
}
CurrentlyOut.defaultProps = {
  rows: [],
  sortOrder: '',
  setColumnSort: () => {},
}

export default CurrentlyOut
