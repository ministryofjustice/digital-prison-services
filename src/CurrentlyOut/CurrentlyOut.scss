.currently-out {

  table-layout: fixed;

  td {
    overflow: hidden;
  }
}