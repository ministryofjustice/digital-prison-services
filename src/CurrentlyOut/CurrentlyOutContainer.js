import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import axios from 'axios'
import { resetError, setLoaded } from '../redux/actions'
import routePaths from '../routePaths'
import Page from '../Components/Page'
import { ASC } from '../tablesorting/sortOrder'
import SortableDataSource from '../tablesorting/SortableDataSource'
import { lastNameComparator } from '../tablesorting/comparatorComposition'
import CurrentlyOut from './CurrentlyOut'

// Data fetchers. One of these is passed to the CurrentlyOut container by a Route in App.js.
// params is the
export const fetchLivingUnitData = params => async () => {
  const response = await axios.get(`/api/movements/livingUnit/${params.livingUnitId}/currently-out`)
  const { currentlyOut, location } = response.data
  return { currentlyOut, location: `Currently out - ${location}` }
}

export const fetchAgencyData = agencyId => async () => {
  const response = await axios.get(`/api/movements/agency/${agencyId}/currently-out`)
  const currentlyOut = response.data
  return { currentlyOut, location: 'Total currently out' }
}

class CurrentlyOutContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      currentlyOut: [],
      location: '',
    }
  }

  componentDidMount() {
    this.getOffendersCurrentlyOut()
  }

  componentDidUpdate(previousProps) {
    const { agencyId, history } = this.props

    if (agencyId !== previousProps.agencyId) {
      history.replace(routePaths.establishmentRoll)
    }
  }

  async getOffendersCurrentlyOut() {
    const { dataFetcher, handleError, resetErrorDispatch, setLoadedDispatch } = this.props

    try {
      resetErrorDispatch()
      setLoadedDispatch(false)

      const nextState = await dataFetcher()

      this.setState(nextState)
      setLoadedDispatch(true)
    } catch (error) {
      handleError(error)
    }
  }

  render() {
    const { currentlyOut, location } = this.state
    return (
      <Page title={location}>
        <SortableDataSource sortOrder={ASC} rows={currentlyOut} comparator={lastNameComparator}>
          <CurrentlyOut />
        </SortableDataSource>
      </Page>
    )
  }
}

CurrentlyOutContainer.propTypes = {
  dataFetcher: PropTypes.func.isRequired,
  handleError: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
  // history from Redux Router Route
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
  }).isRequired,

  // redux state
  agencyId: PropTypes.string.isRequired,

  // redux dispatch
  resetErrorDispatch: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  agencyId: state.app.user.activeCaseLoadId,
  error: state.app.error,
})

const mapDispatchToProps = dispatch => ({
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CurrentlyOutContainer)
