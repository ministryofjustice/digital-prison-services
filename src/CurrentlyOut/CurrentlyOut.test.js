import React from 'react'
import { shallow } from 'enzyme'

import CurrentlyOut from './CurrentlyOut'

describe('Currently out', () => {
  it('should render the page', () => {
    const offenders = [
      {
        offenderNo: 'G000GG',
        firstName: 'FirstName',
        lastName: 'LastName',
        location: 'MDI-1-3-017',
        dateOfBirth: '1960-12-31',
        ipeLevel: 'standard',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
        category: 'C',
        toAgencyDescription: 'To Agency Description',
        commentText: 'Comment text',
      },
    ]
    const wrapper = shallow(<CurrentlyOut sortOrer="ASC" setColumnSort={jest.fn()} rows={offenders} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display toCity when toAgencyDescription is not set ', () => {
    const offenders = [
      {
        offenderNo: 'G000GG',
        firstName: 'FirstName',
        lastName: 'LastName',
        location: 'MDI-1-3-017',
        dateOfBirth: '1960-12-31',
        ipeLevel: 'standard',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
        category: 'C',
        toCity: 'York',
        commentText: 'Comment text',
      },
    ]
    const wrapper = shallow(<CurrentlyOut sortOrer="ASC" setColumnSort={jest.fn()} rows={offenders} />)
    expect(wrapper).toMatchSnapshot()
  })
})
