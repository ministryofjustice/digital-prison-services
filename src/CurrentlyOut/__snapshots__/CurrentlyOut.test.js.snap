// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currently out should display toCity when toAgencyDescription is not set  1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder=""
      />
    </GridCol>
  </GridRow>
  <table
    className="currently-out"
  >
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder=""
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
        <th
          className="straight width10"
        >
          Current location
        </th>
        <th
          className="straight width15"
        >
          Comment
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G000GG"
      >
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G000GG"
          >
            <OffenderImage
              offenderNo="G000GG"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G000GG"
          >
            <OffenderName
              firstName="FirstName"
              lastName="LastName"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          G000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate
            isoDate="1960-12-31"
          />
        </td>
        <td
          className="row-gutters"
        >
          <Connect(Location)
            location="MDI-1-3-017"
          />
        </td>
        <td
          className="row-gutters"
        />
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category="C"
            offenderNo="G000GG"
          />
        </td>
        <td
          className="row-gutters"
        >
          York
        </td>
        <td
          className="row-gutters"
        >
          Comment text
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Currently out should render the page 1`] = `
<Fragment>
  <GridRow>
    <GridCol
      columnFull={false}
      columnOneHalf={false}
      columnOneQuarter={false}
      columnOneThird={false}
      columnThreeQuarters={false}
      columnTwoThirds={false}
      setWidth="one-quarter"
    >
      <SortLov
        setColumnSort={[MockFunction]}
        sortColumn="lastName"
        sortColumns={
          Array [
            "lastName",
          ]
        }
        sortOrder=""
      />
    </GridCol>
  </GridRow>
  <table
    className="currently-out"
  >
    <thead>
      <tr>
        <th
          className="straight width10"
        />
        <th
          className="straight width10"
        >
          <SortableColumn
            column="lastName"
            heading="Name"
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortOrder=""
          />
        </th>
        <th
          className="straight width10"
        >
          Prison no.
        </th>
        <th
          className="straight width10"
        >
          D.O.B.
        </th>
        <th
          className="straight width10"
        >
          Location
        </th>
        <th
          className="straight width10"
        >
          Incentive Level
        </th>
        <th
          className="straight width15"
        >
          Flags
        </th>
        <th
          className="straight width10"
        >
          Current location
        </th>
        <th
          className="straight width15"
        >
          Comment
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="row-gutters"
        key="G000GG"
      >
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G000GG"
          >
            <OffenderImage
              offenderNo="G000GG"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          <OffenderLink
            alerts={false}
            offenderNo="G000GG"
          >
            <OffenderName
              firstName="FirstName"
              lastName="LastName"
            />
          </OffenderLink>
        </td>
        <td
          className="row-gutters"
        >
          G000GG
        </td>
        <td
          className="row-gutters"
        >
          <FormattedDate
            isoDate="1960-12-31"
          />
        </td>
        <td
          className="row-gutters"
        >
          <Connect(Location)
            location="MDI-1-3-017"
          />
        </td>
        <td
          className="row-gutters"
        />
        <td
          className="row-gutters"
        >
          <Flags
            alerts={
              Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ]
            }
            category="C"
            offenderNo="G000GG"
          />
        </td>
        <td
          className="row-gutters"
        >
          To Agency Description
        </td>
        <td
          className="row-gutters"
        >
          Comment text
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
